packages/ra-core/src/controller/usePrevNextController.ts

```diff
@@ -124,37 +125,76 @@ export const usePrevNextController = <RecordType extends RaRecord = any>(
         );
     }
 
-    const [storedParams] = useStore<StoredParams>(
+    const [storedParams] = useStore<Partial<ListParams>>(
         storeKey || `${resource}.listParams`,
         {
-            filter: {},
-            order: sort.order,
-            sort: sort.field,
+            filter: filterDefaultValues,
+            order: initialSort.order,
+            sort: initialSort.field,
         }
     );
 
-    const { data, error, isLoading } = useGetList<RecordType>(
+    const dataProvider = useDataProvider();
+    const queryClient = useQueryClient();
+    const pagination = { page: 1, perPage: limit };
+    const sort = {
+        field: storedParams.sort,
+        order: storedParams.order,
+    };
+    const filter = { ...storedParams.filter, ...permanentFilter };
+    const { meta, ...otherQueryOptions } = queryOptions;
+    const params = { pagination, sort, filter, meta };
+
+    // try to use data from the cache first
+    const queryData = queryClient.getQueryData<{
+        data: RaRecord[];
+        total: number;
+    }>([
         resource,
+        'getList',
         {
-            sort: {
-                field: storedParams.sort,
-                order: storedParams.order,
+            ...params,
+            pagination: {
+                page: storedParams.page,
+                perPage: storedParams.perPage,
             },
-            filter: { ...storedParams.filter, ...filter },
-            pagination: { page: 1, perPage: limit },
         },
-        queryOptions
+    ]);
+    const recordIndexInQueryData = queryData?.data?.findIndex(
+        r => r.id === record?.id
     );
+    const isRecordIndexFirstInNonFirstPage =
+        recordIndexInQueryData === 0 && storedParams.page > 1;
+    const isRecordIndexLastInNonLastPage =
+        recordIndexInQueryData === queryData?.data?.length - 1 &&
+        storedParams.page < queryData?.total / storedParams.perPage;
+    const canUseCacheData =
+        record &&
+        queryData?.data &&
+        recordIndexInQueryData !== -1 &&
+        !isRecordIndexFirstInNonFirstPage &&
+        !isRecordIndexLastInNonLastPage;
 
-    if (!record) return null;
+    // If the previous and next ids are not in the cache, fetch the entire list.
+    // This is necessary e.g. when coming directly to a detail page,
+    // without displaying the list first
+    const { data, error, isLoading } = useQuery(
```

Comment:
I used useQuery instead of useGetList to avoid paying the tax of prepopulating the `getOne` cache, which can be costly for large lists.